File: from-2801-to-2900/2807-insert-greatest-common-divisors-in-linked-list/solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func insertGreatestCommonDivisors(head *ListNode) *ListNode {
    p := head
    var newVal int
    var newNode *ListNode
    for p.Next != nil {
        newVal = gcd(p.Val, p.Next.Val)
        newNode = &ListNode{Val: newVal, Next: p.Next}
        p.Next = newNode
        p = p.Next.Next
    }
    return head
}

func gcd(a, b int) int {
    for a != b {
        if a > b {
            a -= b
        } else {
            b -= a
        }
    }
    return a
}

File: from-2801-to-2900/2807-insert-greatest-common-divisors-in-linked-list/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def insertGreatestCommonDivisors(self, head: Optional[ListNode]) -> Optional[ListNode]:
        def gcd(a: int, b: int) -> int:
            while a != b:
                if a > b:
                    a -= b
                else:
                    b -= a
            return a

        p = head
        while p.next:
            newNode = ListNode(val=gcd(p.val, p.next.val), next=p.next)
            p.next = newNode
            p = p.next.next
        return head